### ISSUE TITLE
OpenJ9 Produces Incorrect Results After Running for a Period of Time

### ISSUE THREAD

**Issue by Qeryu** [eclipse-openj9/openj9#18860]
Java -version output
--------------------

```bash
openjdk version "1.8.0_402-internal"  
OpenJDK Runtime Environment (build 1.8.0_402-internal_2024_01_05_11_05-b00)  
Eclipse OpenJ9 VM (build master-1b60ab828, JRE 1.8.0 Linux amd64-64-Bit Compressed References 20240105_000000 (JIT enabled, AOT enabled)  
OpenJ9   - 1b60ab828  
OMR      - 3a4787401  
JCL      - b058bf977e based on jdk8u402-b05)

javac 1.8.0_402-internal
```

Summary of problem
------------------

When I run the following test case using OpenJ9 JDK8, it results in erroneous output after about 1000 iterations.

```java
class Test {
    int a;
    long b;
    static byte c;
    volatile double dFld;

    int d() {
        int[] e = new int[a];
        long g = 0;
        return (int) g;
    }

    void h() {
        int j = 8;
        short k = 0;
        for (int l = 1; l < 321; l++) {
            k -= b;
            --c;
            d();
            c *= b;
            b = c;
            dFld = j;
        }
        System.out.println(c);
    }

    public static void main(String[] m) {
        Test _instance = new Test();
        for (int f = 0; f < 2000; f++)
            _instance.h();
    }
}
```

Normally, the value of the variable 'c' in all outputs should be 0. However, with OpenJ9 JDK8, some of the outputs show different results. The diff from the normal output is as follows:

```bash
$ diff result_openj9 result_otherjdk
1264,2000c1264,2000
< -15
< -63
< -63
< -63
< -63
< -63
< -63
< -63
< -63
< -63
< -63

...

```

**Comment by Qeryu** [comment]
When running with the `-Xjit:{Test.h(*)*}(optLevel=hot)` option, an incorrect result occurs approximately once in every five runs.

**Comment by pshipton** [comment]
@hzongaro fyi

**Comment by hzongaro** [comment]
I'm seeing failures very infrequently.  The failures I've seen seem to occur if `Test.h` is compiled and run using Dynamic Loop Transfer:

+ (hot) Test.h()V @ 00007F2810B52F74-00007F2810B542FE MethodInProgress - Q_SZ=10 Q_SZI=10 QW=27 j9m=00000000001575
08 bcsz=82 DLT@7 compThreadID=0 CpuLoad=10%(2%avg) JvmCpu=0%

**Comment by hzongaro** [comment]
This could be an issue with PRE.  I am sometimes (perhaps 1/100) able to reproduce the failure with:

```
-Xjit:dltOptLevel=hot,{Test.h*}{hot}\(enableDLTBytecodeIndex=7,lastOptIndex=102,lastOptSubIndex=0\)
```

but not with the following options.  The opt index appears to correspond to PRE.

```
-Xjit:dltOptLevel=hot,{Test.h*}{hot}\(enableDLTBytecodeIndex=7,lastOptIndex=101\)
```

The compilation doesn't manage to complete in time with the `tracePRE` or `traceLastOpt` options, so I tried commenting out most of the tracing code in PRE, and managed to produce some logs from successful compilations showing the trees before and after PRE.  Still looking at them to understand what change might have resulted in the incorrect output.

**Comment by hzongaro** [comment]
I noticed something interesting in the log file, [test.h.optidx102.subidx0.log.txt](https://github.com/eclipse-openj9/openj9/files/14141911/test.h.optidx102.subidx0.log.txt), which I have attached.  Before PRE, we have the following trees (pruned a bit) in `Test.h`:

```
n4281n    BBStart <block_128> (freq 1495) (loop pre-header) (in loop 115)                     [0x7f5b454d09e0] bci=[-1,7,16] rc=0 vc=1465 vn=259 li=-1 udi=- nc=0
n4282n    BBEnd </block_128> =====                                                            [0x7f5b454d0a30] bci=[-1,7,16] rc=0 vc=1142 vn=260 li=-1 udi=- nc=0

n3790n    BBStart <block_82> (freq 9972) (in loop 82)                                         [0x7f5b452a7060] bci=[-1,34,19] rc=0 vc=1467 vn=261 li=-1 udi=- nc=0
n3792n    istore  Test.c B[#426  Static] [flags 0x301 0x0 ]                                   [0x7f5b452a7100] bci=[-1,30,18] rc=0 vc=1467 vn=24 li=3 udi=- nc=1
n3793n      b2i                                                                               [0x7f5b452a7150] bci=[-1,29,18] rc=1 vc=1467 vn=24 li=6 udi=- nc=1
n3794n        i2b                                                                             [0x7f5b452a71a0] bci=[-1,29,18] rc=1 vc=1467 vn=23 li=5 udi=- nc=1
n3795n          iadd                                                                          [0x7f5b452a71f0] bci=[-1,28,18] rc=1 vc=1467 vn=22 li=4 udi=- nc=2
n3796n            iload  Test.c B[#426  Static] [flags 0x301 0x0 ] (cannotOverflow )          [0x7f5b452a7240] bci=[-1,24,18] rc=1 vc=1467 vn=21 li=3 udi=- nc=0 flg=0x1000
n3797n            iconst -1 (X!=0 X<=0 )                                                      [0x7f5b452a7290] bci=[-1,27,18] rc=1 vc=1467 vn=17 li=-1 udi=- nc=0 flg=0x204
   ...
n3034n    BBEnd </block_82> =====                                                             [0x7f5b45508410] bci=[-1,34,19] rc=0 vc=1147 vn=263 li=-1 udi=- nc=0

n3022n    BBStart <block_7> (freq 9974) (in loop 82)                                          [0x7f5b45508050] bci=[-1,7,16] rc=0 vc=1469 vn=264 li=-1 udi=- nc=0
   ...
n45n      istore  Test.c B[#426  Static] [flags 0x301 0x0 ]                                   [0x7f5b45103770] bci=[-1,49,20] rc=0 vc=1469 vn=33 li=3 udi=- nc=1
n44n        b2i                                                                               [0x7f5b45103720] bci=[-1,48,20] rc=1 vc=1469 vn=33 li=13 udi=- nc=1
n43n          i2b                                                                             [0x7f5b451036d0] bci=[-1,48,20] rc=1 vc=1469 vn=32 li=12 udi=- nc=1
n42n            imul                                                                          [0x7f5b45103680] bci=[-1,47,20] rc=1 vc=1469 vn=31 li=11 udi=- nc=2
n3665n            l2i                                                                         [0x7f5b452a4950] bci=[-1,43,20] rc=1 vc=1469 vn=30 li=10 udi=- nc=1
n40n                lloadi  Test.b J[#425  Shadow +8] [flags 0x604 0x0 ] (cannotOverflow )    [0x7f5b451035e0] bci=[-1,43,20] rc=1 vc=1469 vn=29 li=9 udi=- nc=1 flg=0x1000
n39n                  aload  <'this' parm LTest;>[#421  Parm] [flags 0x40000107 0x0 ] (X!=0 X>=0 )  [0x7f5b45103590] bci=[-1,42,20] rc=1 vc=1469 vn=1 li=7 udi=58 nc=0 flg=0x104
n37n              iload  Test.c B[#426  Static] [flags 0x301 0x0 ] (cannotOverflow )          [0x7f5b451034f0] bci=[-1,38,20] rc=1 vc=1469 vn=28 li=3 udi=- nc=0 flg=0x1000
n49n      lstorei  Test.b J[#425  Shadow +8] [flags 0x604 0x0 ]                               [0x7f5b451038b0] bci=[-1,57,21] rc=0 vc=1469 vn=35 li=9 udi=- nc=2
n46n        aload  <'this' parm LTest;>[#421  Parm] [flags 0x40000107 0x0 ] (X!=0 X>=0 )      [0x7f5b451037c0] bci=[-1,52,21] rc=1 vc=1469 vn=1 li=7 udi=59 nc=0 flg=0x104
n48n        i2l                                                                               [0x7f5b45103860] bci=[-1,56,21] rc=1 vc=1469 vn=35 li=14 udi=- nc=1
n47n          iload  Test.c B[#426  Static] [flags 0x301 0x0 ] (cannotOverflow )              [0x7f5b45103810] bci=[-1,53,21] rc=1 vc=1469 vn=34 li=3 udi=- nc=0 flg=0x1000
   ...
n57n      istore  <auto slot 3>[#424  Auto] [flags 0x3 0x0 ]                                  [0x7f5b4517e190] bci=[-1,66,16] rc=0 vc=1469 vn=37 li=-1 udi=4 nc=1
n56n        isub (cannotOverflow )                                                            [0x7f5b4517e140] bci=[-1,66,16] rc=1 vc=1469 vn=37 li=18 udi=- nc=2 flg=0x1000
n54n          iload  <auto slot 3>[#424  Auto] [flags 0x3 0x0 ] (cannotOverflow )             [0x7f5b4517e0a0] bci=[-1,66,16] rc=1 vc=1469 vn=38 li=1 udi=62 nc=0 flg=0x1000
n55n          iconst -1 (X!=0 X<=0 )                                                          [0x7f5b4517e0f0] bci=[-1,66,16] rc=1 vc=1469 vn=17 li=-1 udi=- nc=0 flg=0x204

n3730n    ificmplt --> block_82 BBStart at n3790n ()                                          [0x7f5b452a5da0] bci=[-1,11,16] rc=0 vc=1469 vn=267 li=-1 udi=- nc=2 flg=0x20
n3731n      iload  <auto slot 3>[#424  Auto] [flags 0x3 0x0 ] (cannotOverflow )               [0x7f5b452a5df0] bci=[-1,7,16] rc=1 vc=1469 vn=37 li=1 udi=63 nc=0 flg=0x1000
n3732n      iconst 321 (X!=0 X>=0 )                                                           [0x7f5b452a5e40] bci=[-1,8,16] rc=1 vc=1469 vn=10 li=-1 udi=- nc=0 flg=0x104
n3729n    BBEnd </block_7> =====                                                              [0x7f5b452a5d50] bci=[-1,7,16] rc=0 vc=1155 vn=268 li=-1 udi=- nc=0

n3815n    BBStart <block_83> (freq 0)                                                         [0x7f5b452a7830] bci=[-1,72,24] rc=0 vc=1471 vn=269 li=-1 udi=- nc=0
   ...
n60n          aload  java/lang/System.out Ljava/io/PrintStream;[#430  final Static] [flags 0x20307 0x0 ] (X>=0 )  [0x7f5b4517e280] bci=[-1,72,24] rc=3 vc=1471 vn=39 li=19 udi=- nc=0 flg=0x100
n3608n    astore  <temp slot 7>[#854  Auto] [flags 0x20000007 0x0 ] (X!=0 privatizedInlinerArg )  [0x7f5b452a3780] bci=[-1,72,24] rc=0 vc=1471 vn=39 li=-1 udi=5 nc=1 flg=0x2004
n60n        ==>aload
n3609n    istore  <temp slot 8>[#855  Auto] [flags 0x20000003 0x0 ] (privatizedInlinerArg )   [0x7f5b452a37d0] bci=[-1,75,24] rc=0 vc=1471 vn=42 li=-1 udi=6 nc=1 flg=0x2000
n61n        iload  Test.c B[#426  Static] [flags 0x301 0x0 ] (cannotOverflow )                [0x7f5b4517e2d0] bci=[-1,75,24] rc=1 vc=1471 vn=42 li=3 udi=- nc=0 flg=0x1000
n3621n    ifacmpne --> block_69 BBStart at n3617n (NonoverriddenGuard/VftTest )               [0x7f5b452a3b90] bci=[1,0,735] rc=0 vc=1471 vn=270 li=-1 udi=- nc=2 flg=0x1020
n3619n      aloadi  <vft-symbol>[#346  Shadow] [flags 0x18607 0x0 ] (X!=0 )                   [0x7f5b452a3af0] bci=[-1,72,24] rc=1 vc=1471 vn=44 li=21 udi=- nc=1 flg=0x4
n60n          ==>aload
n3620n      aconst 0xb5e00 (java/io/PrintStream.class) (classPointerConstant X!=0 X>=0 X<=0 )  [0x7f5b452a3b40] bci=[1,0,735] rc=1 vc=1471 vn=43 li=-1 udi=- nc=0 flg=0x10304
n3614n    BBEnd </block_83> =====                                                             [0x7f5b452a3960] bci=[1,2,735] rc=0 vc=1161 vn=271 li=-1 udi=- nc=0
```

Or at a high-level:

```
block_128:  <empty>
block_82:   Test.c = Test.c - 1;
block_7:    Test.c = Test.b*Test.c;
            #424--;
            if (#424 < 321) goto block_82
block_83:   Prepare for System.out.println(#855 = Test.c)
```

After PRE, we have this:

```
n4281n    BBStart <block_128> (freq 1495) (loop pre-header) (in loop 115)                     [0x7f5b454d09e0] bci=[-1,7,16] rc=0 vc=1641 vn=259 li=-1 udi=- nc=0
n4435n    istore  <temp slot 5>[#859  Auto] [flags 0x3 0x0 ]                                  [0x7f5b454d3a00] bci=[-1,53,21] rc=0 vc=1641 vn=654 li=- udi=- nc=1
n4434n      iload  Test.c B[#426  Static] [flags 0x301 0x0 ]                                  [0x7f5b454d39b0] bci=[-1,53,21] rc=1 vc=1641 vn=653 li=3 udi=- nc=0
n4282n    BBEnd </block_128> =====                                                            [0x7f5b454d0a30] bci=[-1,7,16] rc=0 vc=1142 vn=260 li=-1 udi=- nc=0

n3790n    BBStart <block_82> (freq 9972) (in loop 82)                                         [0x7f5b452a7060] bci=[-1,34,19] rc=0 vc=1643 vn=261 li=-1 udi=- nc=0
n3792n    istore  Test.c B[#426  Static] [flags 0x301 0x0 ]                                   [0x7f5b452a7100] bci=[-1,30,18] rc=0 vc=1643 vn=24 li=3 udi=- nc=1
n3793n      b2i                                                                               [0x7f5b452a7150] bci=[-1,29,18] rc=2 vc=1643 vn=24 li=6 udi=- nc=1
n3794n        i2b                                                                             [0x7f5b452a71a0] bci=[-1,29,18] rc=1 vc=1643 vn=23 li=5 udi=- nc=1
n3795n          iadd                                                                          [0x7f5b452a71f0] bci=[-1,28,18] rc=1 vc=1643 vn=22 li=4 udi=- nc=2
n3796n            iload  <temp slot 5>[#859  Auto] [flags 0x3 0x0 ] (cannotOverflow createdByPRE )  [0x7f5b452a7240] bci=[-1,24,18] rc=1 vc=1643 vn=21 li=3 udi=- nc=0 flg=0x41000
n3797n            iconst -1 (X!=0 X<=0 )                                                      [0x7f5b452a7290] bci=[-1,27,18] rc=1 vc=1643 vn=17 li=-1 udi=- nc=0 flg=0x204
n4469n    istore  <temp slot 5>[#859  Auto] [flags 0x3 0x0 ]                                  [0x7f5b454d44a0] bci=[-1,29,18] rc=0 vc=0 vn=688 li=-1 udi=- nc=1
n3793n      ==>b2i
   ...
n3034n    BBEnd </block_82> =====                                                             [0x7f5b45508410] bci=[-1,34,19] rc=0 vc=1147 vn=263 li=-1 udi=- nc=0

n3022n    BBStart <block_7> (freq 9974) (in loop 82)                                          [0x7f5b45508050] bci=[-1,7,16] rc=0 vc=1645 vn=264 li=-1 udi=- nc=0
   ...
n45n      istore  Test.c B[#426  Static] [flags 0x301 0x0 ]                                   [0x7f5b45103770] bci=[-1,49,20] rc=0 vc=1645 vn=33 li=3 udi=- nc=1
n44n        b2i                                                                               [0x7f5b45103720] bci=[-1,48,20] rc=1 vc=1645 vn=33 li=13 udi=- nc=1
n43n          i2b                                                                             [0x7f5b451036d0] bci=[-1,48,20] rc=1 vc=1645 vn=32 li=12 udi=- nc=1
n42n            imul                                                                          [0x7f5b45103680] bci=[-1,47,20] rc=1 vc=1645 vn=31 li=11 udi=- nc=2
n3665n            l2i                                                                         [0x7f5b452a4950] bci=[-1,43,20] rc=1 vc=1645 vn=30 li=10 udi=- nc=1
n40n                lloadi  Test.b J[#425  Shadow +8] [flags 0x604 0x0 ] (cannotOverflow )    [0x7f5b451035e0] bci=[-1,43,20] rc=1 vc=1645 vn=29 li=9 udi=- nc=1 flg=0x1000
n39n                  aload  <'this' parm LTest;>[#421  Parm] [flags 0x40000107 0x0 ] (X!=0 X>=0 )  [0x7f5b45103590] bci=[-1,42,20] rc=1 vc=1645 vn=1 li=7 udi=58 nc=0 flg=0x104
n37n              iload  Test.c B[#426  Static] [flags 0x301 0x0 ] (cannotOverflow )          [0x7f5b451034f0] bci=[-1,38,20] rc=1 vc=1645 vn=28 li=3 udi=- nc=0 flg=0x1000
n49n      lstorei  Test.b J[#425  Shadow +8] [flags 0x604 0x0 ]                               [0x7f5b451038b0] bci=[-1,57,21] rc=0 vc=1645 vn=35 li=9 udi=- nc=2
n46n        aload  <'this' parm LTest;>[#421  Parm] [flags 0x40000107 0x0 ] (X!=0 X>=0 )      [0x7f5b451037c0] bci=[-1,52,21] rc=1 vc=1645 vn=1 li=7 udi=59 nc=0 flg=0x104
n48n        i2l                                                                               [0x7f5b45103860] bci=[-1,56,21] rc=1 vc=1645 vn=35 li=14 udi=- nc=1
n47n          iload  Test.c B[#426  Static] [flags 0x301 0x0 ] (cannotOverflow )              [0x7f5b45103810] bci=[-1,53,21] rc=1 vc=1645 vn=34 li=3 udi=- nc=0 flg=0x1000
   ...
n57n      istore  <auto slot 3>[#424  Auto] [flags 0x3 0x0 ]                                  [0x7f5b4517e190] bci=[-1,66,16] rc=0 vc=1645 vn=37 li=-1 udi=4 nc=1
n56n        isub (cannotOverflow )                                                            [0x7f5b4517e140] bci=[-1,66,16] rc=1 vc=1645 vn=37 li=18 udi=- nc=2 flg=0x1000
n54n          iload  <auto slot 3>[#424  Auto] [flags 0x3 0x0 ] (cannotOverflow )             [0x7f5b4517e0a0] bci=[-1,66,16] rc=1 vc=1645 vn=38 li=1 udi=62 nc=0 flg=0x1000
n55n          iconst -1 (X!=0 X<=0 )                                                          [0x7f5b4517e0f0] bci=[-1,66,16] rc=1 vc=1645 vn=17 li=-1 udi=- nc=0 flg=0x204
n3730n    ificmplt --> block_82 BBStart at n3790n ()                                          [0x7f5b452a5da0] bci=[-1,11,16] rc=0 vc=1645 vn=267 li=-1 udi=- nc=2 flg=0x20
n3731n      iload  <auto slot 3>[#424  Auto] [flags 0x3 0x0 ] (cannotOverflow )               [0x7f5b452a5df0] bci=[-1,7,16] rc=1 vc=1645 vn=37 li=1 udi=63 nc=0 flg=0x1000
n3732n      iconst 321 (X!=0 X>=0 )                                                           [0x7f5b452a5e40] bci=[-1,8,16] rc=1 vc=1645 vn=10 li=-1 udi=- nc=0 flg=0x104
n3729n    BBEnd </block_7> =====                                                              [0x7f5b452a5d50] bci=[-1,7,16] rc=0 vc=1155 vn=268 li=-1 udi=- nc=0

n3815n    BBStart <block_83> (freq 0)                                                         [0x7f5b452a7830] bci=[-1,72,24] rc=0 vc=1647 vn=269 li=-1 udi=- nc=0
   ...
n60n          aload  java/lang/System.out Ljava/io/PrintStream;[#430  final Static] [flags 0x20307 0x0 ] (X>=0 )  [0x7f5b4517e280] bci=[-1,72,24] rc=3 vc=1647 vn=39 li=19 udi=- nc=0 flg=0x100
n3608n    astore  <temp slot 7>[#854  Auto] [flags 0x20000007 0x0 ] (X!=0 privatizedInlinerArg )  [0x7f5b452a3780] bci=[-1,72,24] rc=0 vc=1647 vn=39 li=-1 udi=5 nc=1 flg=0x2004
n60n        ==>aload
n3609n    istore  <temp slot 8>[#855  Auto] [flags 0x20000003 0x0 ] (privatizedInlinerArg )   [0x7f5b452a37d0] bci=[-1,75,24] rc=0 vc=1647 vn=42 li=-1 udi=6 nc=1 flg=0x2000
n61n        iload  <temp slot 5>[#859  Auto] [flags 0x3 0x0 ] (cannotOverflow createdByPRE )  [0x7f5b4517e2d0] bci=[-1,75,24] rc=1 vc=1647 vn=42 li=3 udi=- nc=0 flg=0x41000
```

Or at a high-level, it looks like `#859` has replaced some, but not all updates of `Test.c` in the loop.  Notice in particular that the result of the assignment `Test.c *= Test.b` from block_7 is not used in the decrement in block_83.

```
block_128:  #859 = Test.c;
block_82:   #859 = Test.c = #859 - 1;
block_7:    Test.c = Test.b*Test.c;
            #424--;
            if (#424 < 321) goto block_82
block_83:   Prepare for System.out.println(#855 = #859)
```

@jmesyou, may I ask you to take a look at this?  It's possible this is related to the previously reported problem you're looking at, #18765.

**Comment by hzongaro** [comment]
We haven't investigated this problem enough to get a fix in for the 0.44.0 release. Moving to 0.45.

**Comment by hzongaro** [comment]
I just tried this test with a JDK 17 0.44.0 release build and pre-release build of 0.46.0.  With the 0.44.0 release build, the test produced incorrect results in 6 out of 10000 runs; with the 0.46.0 pre-release build, it produced correct results in 10000 out of 10000 runs, so I believe this was a duplicate of one of the problems that I mentioned earlier.

@Qeryu, I will close this issue as having been fixed.  If you are able to reproduce the problem with a recent build, please reopen the issue.
